include/aspect/particle/output/interface.h

```diff
@@ -0,0 +1,203 @@
+
+#include <aspect/particle/particle.h>
+#include <aspect/particle/definitions.h>
+#include <aspect/plugins.h>
+
+#include <deal.II/base/parameter_handler.h>
+#include <deal.II/base/mpi.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Output
+    {
+      /**
+       * Abstract base class used for classes that generate particle output
+       */
+      template <int dim>
+      class Interface
+      {
+        public:
+          /**
+           * Constructor.
+           */
+          Interface();
+
+          /**
+           * Destructor. Made virtual so that derived classes can be created
+           * and destroyed through pointers to the base class.
+           */
+          virtual ~Interface () {}
+
+
+          /**
+           * Initialization function. This function is called once at the
+           * beginning of the program after parse_parameters is run and after the
+           * SimulatorAccess (if applicable) is initialized.
+           */
+          virtual
+          void
+          initialize ();
+
+          /**
+           * Write data about the particles specified in the first argument to
+           * a file. If possible, encode the current simulation time into this
+           * file using the data provided in the second argument.
+           *
+           * @param [in] particles The set of particles to generate a
+           * graphical representation for
+           * @param [in] data_names The names of the particle properties that
+           * will be written.
+           * @param [in] data_components The number of property components.
+           * Equals one for scalar properties and dim for vector properties,
+           * but any other number is valid as well (e.g. number of compositional
+           * fields).
+           * @param [in] current_time Current time of the simulation, given as
+           * either years or seconds, as selected in the input file. In other
+           * words, output writers do not need to know the units in which time
+           * is described.
+           * @return The name of the file that was written, or
+           * any other information that describes what output was produced if
+           * for example multiple files were created.
+           */
+          virtual
+          std::string
+          output_particle_data(const std::multimap<LevelInd, Particle<dim> > &particles,
+                               const std::vector<std::string>  &data_names,
+                               const std::vector<unsigned int> &data_components,
+                               const double &current_time) = 0;
+
+          /**
+           * Read or write the data of this object for serialization
+           */
+          template <class Archive>
+          void serialize(Archive &ar, const unsigned int version);
+
+          /**
+           * Save the state of the object.
+           */
+          virtual
+          void
+          save (std::ostringstream &os) const;
```

Comment:
I am not sure about the serialize function. As far as I understood, I need to provide a `serialize` function in order for the following to work: `oa << (*this);` or `oa >> (*this);`  and this is what I do in `save` and `load`. But it has been a while since I wrote that, and I do not remember exactly.